File: Data Structures & Algorithms/minimum-stack/submission-2.py
class MinStack:

    def __init__(self):
        self.s = []
        self.minstack = []

    def push(self, val: int) -> None:
        self.s.append(val)
        if self.minstack:
            self.minstack.append(min(val, self.minstack[-1]))
        else:
            self.minstack.append(val)

    def pop(self) -> None:
        if self.s:
            self.s.pop()
            self.minstack.pop()

    def top(self) -> int:
        if self.s:
            return self.s[-1]
        return 0  # or raise IndexError("Stack is empty")

    def getMin(self) -> int:
        if self.minstack:
            return self.minstack[-1]
        return 0  # or raise IndexError("Stack is empty")

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int,int>> s;
        int m = 0;
        for (int i=0;i<heights.size();i++){
            int curr = i;
            while (!s.empty() && s.top().second >= heights[i]) {
                curr = s.top().first;

                int height = s.top().second;
                int width = i - curr;

                m = max(m, height * width);

                s.pop();
            }

            s.push({curr,heights[i]});
        }
        while (!s.empty()) {
            int index = s.top().first;
            int height = s.top().second;
            int width = heights.size() - index;

            m = max(m, height * width);

            s.pop();
        }
        return m;
    }
};
